File: app/src/main/java/com/ash/studios/musify/Adapters/LRAdapter.java
package com.ash.studios.musify.Adapters;

import android.app.Dialog;
import android.content.Context;
import android.media.MediaPlayer;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.ash.studios.musify.Activities.Categories.LRList;
import com.ash.studios.musify.Interfaces.IControl;
import com.ash.studios.musify.Model.Song;
import com.ash.studios.musify.R;
import com.ash.studios.musify.Utils.Instance;
import com.ash.studios.musify.Utils.Utils;
import com.bumptech.glide.Glide;

import java.util.ArrayList;

public class LRAdapter extends RecyclerView.Adapter<LRAdapter.VH> {
    public ArrayList<Song> list;
    private Context context;

    public LRAdapter(Context context, ArrayList<Song> list, ProgressBar pb, TextView nf) {
        this.list = list;
        this.context = context;

        if (pb != null) pb.setVisibility(View.GONE);
        if (nf != null && getItemCount() == 0) nf.setVisibility(View.VISIBLE);
    }

    @NonNull
    @Override
    public LRAdapter.VH onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new LRAdapter.VH(LayoutInflater.from(context).inflate(R.layout.item, parent, false));

    }

    @Override
    public void onBindViewHolder(@NonNull LRAdapter.VH holder, int position) {
        Song song = list.get(position);

        holder.songName.setText(song.getTitle());
        holder.songArtist.setText(song.getArtist());
        holder.duration.setText(Utils.getDuration(song.getDuration()));
        Glide.with(context.getApplicationContext())
                .asBitmap()
                .load(Utils.getAlbumArt(song.getAlbum_id()))
                .placeholder(R.mipmap.icon)
                .into(holder.albumCover);

        holder.itemView.setOnClickListener(v -> {
            Instance.songs = list;
            Instance.position = position;
            ((IControl) context).onStartPlayer();
            Instance.mp.setOnCompletionListener((MediaPlayer.OnCompletionListener) context);
        });

        holder.itemView.setOnLongClickListener(v -> {
            Dialog dialog = Utils.getDialog(context, R.layout.delete_dg);
            TextView title = dialog.findViewById(R.id.del_dg_title);
            TextView body = dialog.findViewById(R.id.del_dg_body);
            TextView cancel = dialog.findViewById(R.id.close_del_dg_btn);
            TextView delete = dialog.findViewById(R.id.del_song_btn);

            title.setText(new StringBuilder("Remove song?"));
            body.setText(new StringBuilder("Selected song will be removed from the Low-Rated."));
            cancel.setOnClickListener(c -> dialog.dismiss());
            delete.setOnClickListener(d -> {
                dialog.dismiss();

                list.remove(position);
                notifyItemRemoved(position);
                notifyItemRangeChanged(position, getItemCount());
                Utils.deleteFromLR(context, song);
                if (getItemCount() == 0) {
                    ((LRList) context).finish();
                    Toast.makeText(context, "No songs in the Low-Rated", Toast.LENGTH_SHORT).show();
                }
            });
            return true;
        });
    }

    @Override
    public int getItemCount() {
        return list == null ? 0 : list.size();
    }

    static class VH extends RecyclerView.ViewHolder {
        TextView songName, songArtist, duration;
        ImageView albumCover;

        public VH(@NonNull View itemView) {
            super(itemView);

            songName = itemView.findViewById(R.id.title);
            songArtist = itemView.findViewById(R.id.artist);
            duration = itemView.findViewById(R.id.duration);
            albumCover = itemView.findViewById(R.id.album_art);
        }
    }
}

File: app/src/main/java/com/ash/studios/musify/adapters/LRAdapter.kt
package com.ash.studios.musify.adapters

import android.content.Context
import android.media.MediaPlayer.OnCompletionListener
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.ash.studios.musify.Interfaces.IService
import com.ash.studios.musify.Models.Song
import com.ash.studios.musify.R
import com.ash.studios.musify.activities.categories.LRList
import com.ash.studios.musify.utils.Instance.mp
import com.ash.studios.musify.utils.Instance.position
import com.ash.studios.musify.utils.Instance.songs
import com.ash.studios.musify.utils.Utils
import com.bumptech.glide.Glide
import com.futuremind.recyclerviewfastscroll.SectionTitleProvider

class LRAdapter(private val context: Context, var list: ArrayList<Song>, pb: ProgressBar?, nf: TextView?) : RecyclerView.Adapter<LRAdapter.VH>(), SectionTitleProvider {
    init {
        if (pb != null) pb.visibility = View.GONE
        if (nf != null && itemCount == 0) nf.visibility = View.VISIBLE
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): VH {
        return VH(LayoutInflater.from(context).inflate(R.layout.item, parent, false))
    }

    override fun onBindViewHolder(holder: VH, p: Int) {
        val song = list[p]
        holder.songName.text = song.title
        holder.songArtist.text = song.artist
        holder.duration.text = Utils.getDuration(song.duration)
        Glide.with(context.applicationContext)
            .asBitmap()
            .load(Utils.getAlbumArt(song.album_id))
            .placeholder(R.drawable.placeholder)
            .into(holder.albumCover)
        holder.itemView.setOnClickListener {
            songs = list
            position = p
            (context as IService).onStartService()
            mp!!.setOnCompletionListener(context as OnCompletionListener)
        }
        holder.itemView.setOnLongClickListener {
            val dialog = Utils.getDialog(context, R.layout.delete_dg)
            val title = dialog.findViewById<TextView>(R.id.del_dg_title)
            val body = dialog.findViewById<TextView>(R.id.del_dg_body)
            val cancel = dialog.findViewById<TextView>(R.id.close_del_dg_btn)
            val delete = dialog.findViewById<TextView>(R.id.del_song_btn)
            title.text = StringBuilder("Remove song?")
            body.text = StringBuilder("Selected song will be removed from the Low-Rated")
            cancel.setOnClickListener { dialog.dismiss() }
            delete.setOnClickListener {
                dialog.dismiss()
                list.removeAt(p)
                notifyItemRemoved(p)
                notifyItemRangeChanged(p, itemCount)
                Utils.deleteFromLR(context, song)
                if (itemCount == 0) {
                    (context as LRList).finish()
                    Toast.makeText(context, "No songs in the Low-Rated", Toast.LENGTH_SHORT).show()
                }
            }
            true
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    override fun getSectionTitle(position: Int): String {
        return list[position].title.substring(0, 1)
    }

    class VH(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var songName: TextView
        var songArtist: TextView
        var duration: TextView
        var albumCover: ImageView

        init {
            songName = itemView.findViewById(R.id.title)
            songArtist = itemView.findViewById(R.id.artist)
            duration = itemView.findViewById(R.id.duration)
            albumCover = itemView.findViewById(R.id.album_art)
        }
    }
}